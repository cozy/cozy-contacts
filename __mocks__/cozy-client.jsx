import React from "react";

const connect = () => WrappedComponent => ({ ...props }) => (
  <WrappedComponent data={[]} fetchStatus={"ready"} {...props} />
);
const withMutation = (whatever, { name }) => WrappedComponent => ({
  ...props
}) => {
  const mockedMutatorProps = { [name]: () => {} };
  return <WrappedComponent {...props} {...mockedMutatorProps} />;
};
const all = () => {};

exports.withMutation = withMutation;
exports.connect = connect;
exports.all = all;
