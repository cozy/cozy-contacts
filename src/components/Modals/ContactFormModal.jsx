import React from "react";
import { PropTypes } from "prop-types";
import { withCreation } from "../../connections/allContacts";
import Modal, { ModalHeader, ModalDescription } from "cozy-ui/react/Modal";
import ContactForm from "../ContactCard/ContactForm";

const ContactFormModal = ({
  onClose,
  title,
  createContact,
  onCreateContact
}) => (
  <Modal
    overflowHidden={true}
    dismissAction={onClose}
    into="body"
    size="xlarge"
  >
    <ModalHeader>{title}</ModalHeader>
    <ModalDescription className="u-mt-half">
      <ContactForm
        onSubmit={contact =>
          createContact({ ...contact, _type: "io.cozy.contacts" }).then(resp =>
            onCreateContact(resp.data[0])
          )
        }
        onCancel={onClose}
      />
    </ModalDescription>
  </Modal>
);
ContactFormModal.propTypes = {
  onClose: PropTypes.func.isRequired,
  title: PropTypes.string.isRequired,
  createContact: PropTypes.func.isRequired,
  onCreateContact: PropTypes.func.isRequired
};

export default withCreation(ContactFormModal);
