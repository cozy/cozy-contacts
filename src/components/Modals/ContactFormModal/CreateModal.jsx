import React from 'react'
import { useNavigate } from 'react-router-dom'

import { useClient, useQueryAll } from 'cozy-client'
import AddModal from 'cozy-ui/transpiled/react/Contacts/AddModal'
import { useSelectedGroup } from 'cozy-ui/transpiled/react/Contacts/GroupsSelect/GroupsSelectProvider'

import { createOrUpdateContact } from '@/connections/allContacts'
import { buildContactsQueryByFamilyNameGivenNameEmailCozyUrl } from '@/queries/queries'

const CreateModal = () => {
  const navigate = useNavigate()
  const client = useClient()
  const { selectedGroup } = useSelectedGroup()

  const contactsQueryByFamilyNameGivenNameEmailCozyUrl =
    buildContactsQueryByFamilyNameGivenNameEmailCozyUrl()

  const contacts = useQueryAll(
    contactsQueryByFamilyNameGivenNameEmailCozyUrl.definition,
    contactsQueryByFamilyNameGivenNameEmailCozyUrl.options
  )

  const onSubmit = async formData =>
    await createOrUpdateContact({
      client,
      formData,
      selectedGroup
    })

  const onClose = () => navigate('/')

  return <AddModal contacts={contacts} onSubmit={onSubmit} onClose={onClose} />
}

export default CreateModal
