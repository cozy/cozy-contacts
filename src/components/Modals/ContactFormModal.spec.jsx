import React from 'react'
import { render, fireEvent } from '@testing-library/react'

import { createMockClient } from 'cozy-client'

import { DumbContactFormModal } from './ContactFormModal'
import AppLike from '../../tests/Applike'

const client = createMockClient({})

describe('ContactFormModal component', () => {
  let props

  beforeEach(() => {
    props = {
      afterMutation: jest.fn(),
      contact: null,
      createContact: jest.fn().mockResolvedValue({ data: 'created' }),
      onClose: jest.fn(),
      title: 'Edit contact',
      updateContact: jest.fn().mockResolvedValue({ data: 'updated' }),
      t: jest.fn(x => x)
    }
  })

  it('should render a contact form in a modal', () => {
    const contact = {
      name: {
        familyName: 'John',
        givenName: 'Doe'
      }
    }
    const propsWithContact = {
      ...props,
      contact
    }

    const jsx = (
      <AppLike client={client}>
        <DumbContactFormModal {...propsWithContact} />
      </AppLike>
    )

    const { getByRole } = render(jsx)
    const firstNameInput = getByRole('textbox', { name: 'Firstname' })
    const lastNameInput = getByRole('textbox', { name: 'Lastname' })
    expect(firstNameInput.value).toBe('Doe')
    expect(lastNameInput.value).toBe('John')
  })

  it('should pass a new contact to the creation function', () => {
    const formData = {
      company: 'Cozy Cloud'
    }

    const jsx = (
      <AppLike client={client}>
        <DumbContactFormModal {...props} />
      </AppLike>
    )

    const expected = {
      address: [],
      birthday: undefined,
      company: 'Cozy Cloud',
      cozy: [],
      displayName: '',
      email: [],
      fullname: '',
      indexes: { byFamilyNameGivenNameEmailCozyUrl: null },
      jobTitle: '',
      metadata: { cozy: true, version: 1 },
      name: { familyName: undefined, givenName: undefined },
      note: '',
      phone: [],
      relationships: { groups: { data: [] } }
    }

    const { getByRole } = render(jsx)
    const companyInput = getByRole('textbox', { name: 'Company' })
    fireEvent.change(companyInput, {
      target: { value: formData.company }
    })
    const submitButton = getByRole('button', { name: 'save' })
    fireEvent.click(submitButton)

    expect(props.createContact).toHaveBeenCalledWith(expected)
  })

  it('should pass previous contact data to the update function', () => {
    const contact = {
      name: {
        familyName: 'John',
        givenName: 'Doe'
      }
    }
    const propsWithContact = {
      ...props,
      contact
    }
    const formData = {
      company: 'Cozy Cloud'
    }

    const expected = {
      address: [],
      birthday: undefined,
      company: 'Cozy Cloud',
      cozy: [],
      displayName: 'Doe John',
      email: [],
      fullname: 'Doe John',
      indexes: { byFamilyNameGivenNameEmailCozyUrl: 'johndoe' },
      jobTitle: '',
      metadata: { cozy: true, version: 1 },
      name: {
        familyName: 'John',
        givenName: 'Doe'
      },
      note: '',
      phone: [],
      relationships: { groups: { data: [] } }
    }

    const jsx = (
      <AppLike client={client}>
        <DumbContactFormModal {...propsWithContact} />
      </AppLike>
    )

    const { getByRole } = render(jsx)
    const companyInput = getByRole('textbox', { name: 'Company' })
    fireEvent.change(companyInput, {
      target: { value: formData.company }
    })
    const submitButton = getByRole('button', { name: 'save' })
    fireEvent.click(submitButton)

    expect(props.updateContact).toHaveBeenCalledWith(expected)
  })
})
