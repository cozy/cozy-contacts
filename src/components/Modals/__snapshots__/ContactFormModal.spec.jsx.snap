// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactFormModal component should render a contact form in a modal 1`] = `
<Modal
  closable={true}
  dismissAction={[MockFunction]}
  into="body"
  mobileFullscreen={true}
  overflowHidden={true}
  primaryType="regular"
  secondaryType="secondary"
  size="xlarge"
>
  <ModalHeader>
    Edit contact
  </ModalHeader>
  <ModalContent>
    <withI18n(ContactForm)
      contact={
        Object {
          "birthday": "1959-05-15",
          "name": Object {
            "familyName": "John",
            "givenName": "Doe",
          },
        }
      }
      onSubmit={[Function]}
    />
  </ModalContent>
  <ModalFooter
    className="u-ta-right"
  >
    <DefaultButton
      label="cancel"
      onClick={[MockFunction]}
      theme="secondary"
      type="button"
    />
    <DefaultButton
      busy={false}
      label="save"
      onClick={[Function]}
      type="submit"
    />
  </ModalFooter>
</Modal>
`;
